frame/ranked-collective/src/lib.rs

```diff
@@ -0,0 +1,539 @@
+	}
+	fn ayes(&self, _: Rank) -> Votes {
+		self.bare_ayes
+	}
+	fn support(&self, class: Rank) -> Perbill {
+		Perbill::from_rational(self.bare_ayes, M::get_max_voters(class))
+	}
+	fn approval(&self, _: Rank) -> Perbill {
+		Perbill::from_rational(self.ayes, 1.max(self.ayes + self.nays))
+	}
+	#[cfg(feature = "runtime-benchmarks")]
+	fn unanimity(class: Rank) -> Self {
+		Self {
+			bare_ayes: M::get_max_voters(class),
+			ayes: M::get_max_voters(class),
+			nays: 0,
+			dummy: PhantomData,
+		}
+	}
+	#[cfg(feature = "runtime-benchmarks")]
+	fn rejection(class: Rank) -> Self {
+		Self { bare_ayes: 0, ayes: 0, nays: M::get_max_voters(class), dummy: PhantomData }
+	}
+	#[cfg(feature = "runtime-benchmarks")]
+	fn from_requirements(support: Perbill, approval: Perbill, class: Rank) -> Self {
+		let c = M::get_max_voters(class);
+		let ayes = support * c;
+		let nays = ((ayes as u64) * 1_000_000_000u64 / approval.deconstruct() as u64) as u32 - ayes;
+		Self { bare_ayes: ayes, ayes, nays, dummy: PhantomData }
+	}
+}
+
+/// Record needed for every member.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct MemberRecord {
+	/// The rank of the member.
+	rank: Rank,
+}
+
+/// Record needed for every vote.
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum VoteRecord {
+	/// Vote was an aye with given vote weight.
+	Aye(Votes),
+	/// Vote was a nay with given vote weight.
+	Nay(Votes),
+}
+
+impl From<(bool, Votes)> for VoteRecord {
+	fn from((aye, votes): (bool, Votes)) -> Self {
+		match aye {
+			true => VoteRecord::Aye(votes),
+			false => VoteRecord::Nay(votes),
+		}
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote regardless of rank.
+pub struct Unit;
+impl Convert<Rank, Votes> for Unit {
+	fn convert(_: Rank) -> Votes {
+		1
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote plus an additional vote for every excess rank
+/// they have. I.e.:
+///
+/// - Each member with no excess rank gets 1 vote;
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 2 votes;
+/// - ...with an excess rank of 3 gets 3 votes;
+/// - ...with an excess rank of 4 gets 4 votes.
+pub struct Linear;
+impl Convert<Rank, Votes> for Linear {
+	fn convert(r: Rank) -> Votes {
+		(r + 1) as Votes
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote plus additional votes for every excess rank
+/// they have incrementing by one vote for each excess rank. I.e.:
+///
+/// - Each member with no excess rank gets 1 vote;
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 3 votes;
+/// - ...with an excess rank of 3 gets 6 votes;
+/// - ...with an excess rank of 4 gets 10 votes.
+pub struct Geometric;
+impl Convert<Rank, Votes> for Geometric {
+	fn convert(r: Rank) -> Votes {
+		let v = (r + 1) as Votes;
+		v * (v + 1) / 2
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
```

Comment:
Do we also need a `#[pallet::origin]` here and a corresponding `EnsureCollective` check specifically for this pallet?